### PR TITLE
Fix unexpected replacement of the whole volume object

### PR DESCRIPTION
**Pull request by weihanglo**
**Problem:**

Fix unexpected replacement of the whole volume object while attaching volume to backup

**Solution:**

The `Update` method actually is an HTTP `PUT` method, of which the semantic is replacing the whole object. Since `VolumeSpec` lacks some field like `Migratable`, this leads `Spec.Migratable` to update
to bool's default value `false`.

Then, if the volume is in RWX and Migratable mode, the Longhorn share manager controller will try syncing its state[^1], which will finally result in a detachment[^2].

~~To solve this, we use `Patch` instead of `Update` (PUT).~~

The root cause is we use Longhorn v1.1.1 in our helm chart but the client code version in harvester didn't get an upgrade. Therefore the `VolumeSpec.Migratable` field is missing. Upgrade Longhorn to v1.1.1 can solve this.

[^1]: https://github.com/longhorn/longhorn-manager/blob/v1.1.1/controller/share_manager_controller.go#L151-L158
[^2]: https://github.com/longhorn/longhorn-manager/blob/v1.1.1/controller/share_manager_controller.go#L416-L422

<details>

<summary> 👁 Expand to read the wrong request payload to <code>PUT</code> a volume </summary>

```json
{
  "kind": "Volume",
  "apiVersion": "longhorn.io/v1beta1",
  "metadata": {
    "name": "pvc-397acf57-d073-41c7-938a-e5c56f186aa3",
    "namespace": "longhorn-system",
    "uid": "59c602c4-2b4f-472d-a350-71ea312157aa",
    "resourceVersion": "79519",
    "generation": 3,
    "creationTimestamp": "2021-06-24T08:51:53Z",
    "labels": {
      "longhornvolume": "pvc-397acf57-d073-41c7-938a-e5c56f186aa3"
    },
    "finalizers": [
      "longhorn.io"
    ],
    "managedFields": [
      {
        "manager": "longhorn-manager",
        "operation": "Update",
        "apiVersion": "longhorn.io/v1beta1",
        "time": "2021-06-24T08:51:53Z",
        "fieldsType": "FieldsV1",
        "fieldsV1": {
          "f:metadata": {
            "f:finalizers": {
              ".": {},
              "v:\"longhorn.io\"": {}
            },
            "f:labels": {
              ".": {},
              "f:longhornvolume": {}
            }
          },
          "f:spec": {},
          "f:status": {}
        }
      }
    ]
  },
  "spec": {
    "size": "10737418240",
    "frontend": "blockdev",
    "fromBackup": "",
    "numberOfReplicas": 3,
    "dataLocality": "disabled",
    "staleReplicaTimeout": 30,
    "nodeID": "h0",
    "engineImage": "longhornio/longhorn-engine:v1.1.1",
    "recurringJobs": null,
    "baseImage": "",
    "Standby": false,
    "diskSelector": null,
    "nodeSelector": null,
    "disableFrontend": false,
    "revisionCounterDisabled": false,
    "lastAttachedBy": "",
    "accessMode": "rwx"
  },
  "status": {
    "ownerID": "h0",
    "state": "detached",
    "robustness": "unknown",
    "currentNodeID": "",
    "currentImage": "longhornio/longhorn-engine:v1.1.1",
    "kubernetesStatus": {
      "pvName": "pvc-397acf57-d073-41c7-938a-e5c56f186aa3",
      "pvStatus": "Bound",
      "namespace": "default",
      "pvcName": "bb-1-rootdisk-mklow",
      "lastPVCRefAt": "",
      "workloadsStatus": [
        {
          "podName": "virt-launcher-bb-1-vjmsd",
          "podStatus": "Running",
          "workloadName": "bb-1",
          "workloadType": "VirtualMachineInstance"
        }
      ],
      "lastPodRefAt": "2021-06-24T08:59:40Z"
    },
    "conditions": {
      "restore": {
        "type": "restore",
        "status": "False",
        "lastProbeTime": "",
        "lastTransitionTime": "2021-06-24T08:52:02Z",
        "reason": "",
        "message": ""
      },
      "scheduled": {
        "type": "scheduled",
        "status": "True",
        "lastProbeTime": "",
        "lastTransitionTime": "2021-06-24T09:00:30Z",
        "reason": "",
        "message": ""
      },
      "toomanysnapshots": {
        "type": "toomanysnapshots",
        "status": "False",
        "lastProbeTime": "",
        "lastTransitionTime": "2021-06-24T08:52:02Z",
        "reason": "",
        "message": ""
      }
    },
    "lastBackup": "",
    "lastBackupAt": "",
    "pendingNodeID": "",
    "frontendDisabled": false,
    "restoreRequired": false,
    "restoreInitiated": false,
    "remountRequestedAt": "",
    "expansionRequired": false,
    "isStandby": false,
    "actualSize": 0,
    "lastDegradedAt": "",
    "shareEndpoint": "",
    "shareState": ""
  }
}
```

</details>

**Related Issue:**

#932 

**Test plan:**

Same as #932.

### CODE REVIEW

File: pkg/controller/master/backup/backupcontent.go [hunk 1]
```diff
@@ -340,14 +349,37 @@ func (h *ContentHandler) reconcileLonghornVolumes(vm *kubevirtv1.VirtualMachine)
...
+			//
+			// The reason choosing JSON patch[1] is that for current Longhorn
+			// `VolumeSpec`, there is no field get a `omitempty` tag, so field
+			// like `Spec.Size` will serialize to 0. Using JSON-merge patch[2]
+			// may lead to apply unexpected default value. Also, CRD does not
+			// support strategy patch[3].
+			//
+			// [1]: https://datatracker.ietf.org/doc/html/rfc6902
+			// [2]: https://datatracker.ietf.org/doc/html/rfc7386
+			// [3]: https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/#advanced-features-and-flexibility
```
Comment (guangbochen) [inline pkg/controller/master/backup/backupcontent.go]: in what case the `spec.migratable` is equal to false, based on my testing this value always exists and defaults to `true`. 
Additionally, we can check with the LH team why this field is missing and we may set it to default to true if this field is nil, and then we can call the update method directly?

Comment (weihanglo) [reply pkg/controller/master/backup/backupcontent.go]: Here is the payload of our HTTP PUT payload  generated by `Update` method.

<details>
<summary>expand</summary>

```json
{
  "kind": "Volume",
  "apiVersion": "longhorn.io/v1beta1",
  "metadata": {
    "name": "pvc-397acf57-d073-41c7-938a-e5c56f186aa3",
    "namespace": "longhorn-system",
    "uid": "59c602c4-2b4f-472d-a350-71ea312157aa",
    "resourceVersion": "79519",
    "generation": 3,
    "creationTimestamp": "2021-06-24T08:51:53Z",
    "labels": {
      "longhornvolume": "pvc-397acf57-d073-41c7-938a-e5c56f186aa3"
    },
    "finalizers": [
      "longhorn.io"
    ],
    "managedFields": [
      {
        "manager": "longhorn-manager",
        "operation": "Update",
        "apiVersion": "longhorn.io/v1beta1",
        "time": "2021-06-24T08:51:53Z",
        "fieldsType": "FieldsV1",
        "fieldsV1": {
          "f:metadata": {
            "f:finalizers": {
              ".": {},
              "v:\"longhorn.io\"": {}
            },
            "f:labels": {
              ".": {},
              "f:longhornvolume": {}
            }
          },
          "f:spec": {},
          "f:status": {}
        }
      }
    ]
  },
  "spec": {
    "size": "10737418240",
    "frontend": "blockdev",
    "fromBackup": "",
    "numberOfReplicas": 3,
    "dataLocality": "disabled",
    "staleReplicaTimeout": 30,
    "nodeID": "h0",
    "engineImage": "longhornio/longhorn-engine:v1.1.1",
    "recurringJobs": null,
    "baseImage": "",
    "Standby": false,
    "diskSelector": null,
    "nodeSelector": null,
    "disableFrontend": false,
    "revisionCounterDisabled": false,
    "lastAttachedBy": "",
    "accessMode": "rwx"
  },
  "status": {
    "ownerID": "h0",
    "state": "detached",
    "robustness": "unknown",
    "currentNodeID": "",
    "currentImage": "longhornio/longhorn-engine:v1.1.1",
    "kubernetesStatus": {
      "pvName": "pvc-397acf57-d073-41c7-938a-e5c56f186aa3",
      "pvStatus": "Bound",
      "namespace": "default",
      "pvcName": "bb-1-rootdisk-mklow",
      "lastPVCRefAt": "",
      "workloadsStatus": [
        {
          "podName": "virt-launcher-bb-1-vjmsd",
          "podStatus": "Running",
          "workloadName": "bb-1",
          "workloadType": "VirtualMachineInstance"
        }
      ],
      "lastPodRefAt": "2021-06-24T08:59:40Z"
    },
    "conditions": {
      "restore": {
        "type": "restore",
        "status": "False",
        "lastProbeTime": "",
        "lastTransitionTime": "2021-06-24T08:52:02Z",
        "reason": "",
        "message": ""
      },
      "scheduled": {
        "type": "scheduled",
        "status": "True",
        "lastProbeTime": "",
        "lastTransitionTime": "2021-06-24T09:00:30Z",
        "reason": "",
        "message": ""
      },
      "toomanysnapshots": {
        "type": "toomanysnapshots",
        "status": "False",
        "lastProbeTime": "",
        "lastTransitionTime": "2021-06-24T08:52:02Z",
        "reason": "",
        "message": ""
      }
    },
    "lastBackup": "",
    "lastBackupAt": "",
    "pendingNodeID": "",
    "frontendDisabled": false,
    "restoreRequired": false,
    "restoreInitiated": false,
    "remountRequestedAt": "",
    "expansionRequired": false,
    "isStandby": false,
    "actualSize": 0,
    "lastDegradedAt": "",
    "shareEndpoint": "",
    "shareState": ""
  }
}
```
</details>

I think the root cause is [Longhorn v1.1.0 lacks Migratable field][1], so a PUT (replace) operation will lead `Migratable` to set to false by default.

[1]: https://github.com/longhorn/longhorn-manager/blob/v1.1.0/types/resource.go#L76-L94

Comment (weihanglo) [reply pkg/controller/master/backup/backupcontent.go]: Wait that's weird. We're using 1.1.0 why we need the Migratable field.

Comment (weihanglo) [reply pkg/controller/master/backup/backupcontent.go]: Maybe we just need an update to v1.1.1, since our helm chart already uses Longhorn v1.1.1

Comment (guangbochen) [reply pkg/controller/master/backup/backupcontent.go]: https://github.com/harvester/harvester/blob/v0.2/deploy/charts/harvester/Chart.yaml#L56 it is already using 1.1.1, not sure how u get the 1.1.0?

Comment (weihanglo) [reply pkg/controller/master/backup/backupcontent.go]: Our go.mod still uses v1.1.0. Both v0.2 and master branch

https://github.com/harvester/harvester/blob/v0.2/go.mod#L60
https://github.com/harvester/harvester/blob/158a35d/go.mod#L68

Comment (weihanglo) [reply pkg/controller/master/backup/backupcontent.go]: If it's applicable, I'll try upgrading Longhorn dependency to v1.1.1

Comment (weihanglo) [reply pkg/controller/master/backup/backupcontent.go]: PR Updated. PTAL thanks!

**Review by guangbochen — APPROVED** [review]
LGTM

**Pull request merged**